applications/OptimizationApplication/custom_utilities/response/linear_strain_energy_response_utils.cpp

```diff
@@ -0,0 +1,376 @@
+    });
+
+    const double local_condition_value = block_for_each<SumReduction<double>>(rModelPart.Conditions(), tls_type(), [&](auto& rCondition, tls_type& rTLS) {
+        Matrix& r_lhs = std::get<0>(rTLS);
+        Vector& r_rhs = std::get<1>(rTLS);
+        Vector& r_u = std::get<2>(rTLS);
+        return CalculateEntityStrainEnergy(rCondition, r_lhs, r_rhs, r_u, rModelPart.GetProcessInfo());
+    });
+
+    return rModelPart.GetCommunicator().GetDataCommunicator().SumAll(local_element_value + local_condition_value);
+
+    KRATOS_CATCH("");
+}
+
+void LinearStrainEnergyResponseUtils::CalculateSensitivity(
+    const std::vector<ModelPart*>& rEvaluatedModelParts,
+    const SensitivityModelPartVariablesListMap& rSensitivityModelPartVariableInfo,
+    const double PerturbationSize)
+{
+    KRATOS_TRY
+
+    ResponseUtils::CheckAndPrepareModelPartsForSensitivityComputation(rEvaluatedModelParts, rSensitivityModelPartVariableInfo, SELECTED, {&SHAPE_SENSITIVITY});
+
+    // calculate sensitivities for each and every model part w.r.t. their sensitivity variables list
+    for (const auto& it : rSensitivityModelPartVariableInfo) {
+        auto& r_sensitivity_model_part = *(it.first);
+        for (auto& r_variable : it.second) {
+            std::visit([&](auto&& r_variable) {
+                if (*r_variable == YOUNG_MODULUS_SENSITIVITY) {
+                    CalculateStrainEnergyLinearlyDependentPropertySensitivity(r_sensitivity_model_part, YOUNG_MODULUS, YOUNG_MODULUS_SENSITIVITY);
+                } else if (*r_variable == THICKNESS_SENSITIVITY) {
+                    CalculateStrainEnergyLinearlyDependentPropertySensitivity(r_sensitivity_model_part, THICKNESS, THICKNESS_SENSITIVITY);
+                } else if (*r_variable == POISSON_RATIO_SENSITIVITY) {
+                    CalculateStrainEnergyFiniteDifferencePropertySensitivity(r_sensitivity_model_part, PerturbationSize, POISSON_RATIO, POISSON_RATIO_SENSITIVITY);
+                } else if (*r_variable == SHAPE_SENSITIVITY) {
+                    CalculateStrainEnergyShapeSensitivity(r_sensitivity_model_part, PerturbationSize, SHAPE_SENSITIVITY);
+                } else {
+                    KRATOS_ERROR
+                        << "Unsupported sensitivity w.r.t. " << r_variable->Name()
+                        << " requested for " << r_sensitivity_model_part.FullName()
+                        << ". Followings are supported sensitivity variables:"
+                        << "\n\t" << YOUNG_MODULUS_SENSITIVITY.Name()
+                        << "\n\t" << THICKNESS_SENSITIVITY.Name()
+                        << "\n\t" << POISSON_RATIO_SENSITIVITY.Name()
+                        << "\n\t" << SHAPE_SENSITIVITY.Name();
+                }
+            }, r_variable);
+        }
+    }
+
+    KRATOS_CATCH("");
+}
+
+template<class TEntityType>
+void LinearStrainEnergyResponseUtils::CalculateStrainEnergyEntityShapeSensitivity(
```

Comment:
To make it more elaborate, I suggest renaming it to CalculateEntityResidualShapeSensitivty

Reply:
changed to `CalculateStrainEnergySemiAnalyticShapeSensitivity`